src/frontend/src/expr/type_inference/cast.rs

```diff
@@ -114,76 +116,145 @@ pub fn align_array_and_element(
     Ok(array_type)
 }
 
+/// Returns `Ok` if `ok` is true, otherwise returns a placeholder [`CastError`] to be further
+/// wrapped with a more informative context in [`cast`].
+fn canmeh(ok: bool) -> CastResult {
+    if ok {
+        Ok(())
+    } else {
+        bail_cast_error!()
```

Comment:
A little confused how it works. So this is an empty leaf CastError? Why it's not shown in the err msg stack? (Why we need it?)

Reply:
Yes. In error report we'll disregard all empty and repeated entries, so it won't appear.

To preserve the reason for a cast failure, we must return a `Result` in functions like `cast_struct`. However, the default error message "cannot cast" is uniformly applied in `cast`, so we simply return an empty leaf error and let the caller to fill the context.

Reply:
It sounds like `Result<(), Option<CastError>>` 🤪